### PR TITLE
feat(lsp): inlay hints

### LINKED ISSUE #18086: LSP: support InlayHint, a new feature in LSP Spec 3.17
### Feature description

[The inlay hint feature](https://github.com/microsoft/language-server-protocol/issues/956) is stabilized in latest [LSP Spec 3.17](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/), and implemented in VSCode.

The latest stable version of rust-analyzer 2022-04-11 already changed the inlay hint interface from the experimental one to the finalized one. Thus it no longer work with nvim's builtin LSP and [lsp_extensions.nvim](https://github.com/nvim-lua/lsp_extensions.nvim). Since it's not an experimental feature anymore, I think it's better to be implemented inside nvim rather than plugins.

### PR DESCRIPTION
**Pull request by dchinmay2**
Implements inlay hints by adding automatic refresh and a public interface on top of #23736

Closes #18086

### CODE REVIEW

File: runtime/lua/vim/lsp/_inlay_hint.lua
Comment (glepnir) [inline runtime/lua/vim/lsp/_inlay_hint.lua:197]: not sure is the good way create timer for each buffer ,  and in `on_lines`  callback has `changedtick` param you can use  pass to request like `make_request(cb_bufnr, changedtick)`,  and when  lsp request callback running check `nvim_get_buf_chagnetick` make a compare. `if before_changetick < nvim_get_buf_changetick` do defer reqeust again.   some undo/redo also increase `changetick` and also some change without text chang also increase `changtick` .  this is my thought  and I do it on documentsymbol request . it combine 1000ms change to one request . 
 https://github.com/nvimdev/lspsaga.nvim/blob/3546dca5560b6868612acc567f6f3c07cf61c076/lua/lspsaga/symbol/init.lua#L25

Comment (p00f) [reply runtime/lua/vim/lsp/_inlay_hint.lua:197]: I don't like this either, is there a better way of remembering which buffers have inlay hint enabled (to handle refresh request)?

File: runtime/lua/vim/lsp/_inlay_hint.lua
Comment (jdrouhard) [inline runtime/lua/vim/lsp/_inlay_hint.lua:192]: You'll also want handlers for `on_reload` and `on_detach`. `on_reload` should clear the cache and namespace, and refire a request. `on_detach` should clear out any cache related to that buffer or there will be a memory leak as more buffers accumulate state.

You may also want an `LspDetach` autocmd that does the same thing (in case LSP is detached from a buffer manually).

File: runtime/lua/vim/lsp/_inlay_hint.lua
Comment (jdrouhard) [inline runtime/lua/vim/lsp/_inlay_hint.lua:20]: is this used outside this module any more?

File: runtime/lua/vim/lsp/_inlay_hint.lua
Comment (p00f) [reply runtime/lua/vim/lsp/_inlay_hint.lua]: done

File: runtime/lua/vim/lsp/_inlay_hint.lua
Comment (p00f) [reply runtime/lua/vim/lsp/_inlay_hint.lua:194]: done

File: runtime/lua/vim/lsp/_inlay_hint.lua
Comment (p00f) [reply runtime/lua/vim/lsp/_inlay_hint.lua]: done

File: runtime/lua/vim/lsp/_inlay_hint.lua
Comment (p00f) [reply runtime/lua/vim/lsp/_inlay_hint.lua:199]: That is what I was doing, except how do you cancel pending requests?

File: runtime/lua/vim/lsp/_inlay_hint.lua
Comment (p00f) [reply runtime/lua/vim/lsp/_inlay_hint.lua]: Will do

File: runtime/doc/lsp.txt
Comment (lewis6991) [inline runtime/doc/lsp.txt:719]: Do we really need 4 functions for this?

Comment (p00f) [reply runtime/doc/lsp.txt:719]: Do you want only a toggle function?

Comment (p00f) [reply runtime/doc/lsp.txt:719]: Aren't enable and disable necessary for scripting

Comment (p00f) [reply runtime/doc/lsp.txt:719]: And toggle is necessary if you want a mapping to toggle

File: runtime/doc/lsp.txt
Comment (justinmk) [inline runtime/doc/lsp.txt:702]: Why doesn't this live in `vim.lsp.buf.*` like almost all other buffer-relative LSP features? Strange that it has a `buf_` prefix instead.

File: runtime/doc/lsp.txt
Comment (justinmk) [inline runtime/doc/lsp.txt:708]: - "enable" can take an `enable` parameter, then you don't need the `_disable` variant.
- won't be long before someone asks for a getter (`is_enabled`), right? So if we add that now, then we also don't need `toggle()`.

**Comment by justinmk** [comment]
When I suggested skipping the public API in https://github.com/neovim/neovim/pull/23736 , my thinking was: just enable it by default. It doesn't need to be configurable.

Does LSP actually require the enable/disable stuff to be configured client-side? Seems like this could/should be a setting that could be sent to the server, using a standard "configure" LSP request. Does LSP have something like that?

File: runtime/doc/news.txt
Comment (justinmk) [inline runtime/doc/news.txt:87]: news.txt (at minimum) is a great place to link to reference URLs such as https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_inlayHint

**Comment by clason** [comment]
> When I suggested skipping the public API in #23736 , my thinking was: just enable it by default. It doesn't need to be configurable.
> 
> Does LSP actually require the enable/disable stuff to be configured client-side? Seems like this could/should be a setting that could be sent to the server, using a standard "configure" LSP request. Does LSP have something like that? Can the client-side "capabilities" map be used to enable/disable inlay hints?

Inlay hints can be quite obtrusive (even moreso than diagnostics); they should definitely be under the user's control. (I refer to the semantic tokens decision, which triggered a deluge of "my highlight is suddenly broken!!eleven!" issues for nvim-treesitter.)

**Comment by justinmk** [comment]
> Inlay hints can be quite obtrusive (even moreso than diagnostics); they should definitely be under the user's control. (I refer to the semantic tokens decision,

- Does the "client capabilities" give control or not?
- I don't see any other `enable()` methods in `:help lsp`. Why is this the only one?

**Comment by clason** [comment]
> Does the "client capabilities" give control or not?

Not enough; that is all-or-nothing. The ask here is to only view the inlay hints on demand. not disable them completely

> I don't see any other enable() methods in :help lsp. Why is this the only one?

Because we

1.  started with the simple, non-controversial features;
2. made mistakes for some of the other ones.

**Comment by justinmk** [comment]
> > I don't see any other enable() methods in :help lsp. Why is this the only one?
> 
> Because we (1) started with the simple, non-controversial features ... (2) made mistakes for some of the other ones.

To avoid more mistakes can we look at `vim.diagnostic.config`? Instead of adding lots of `foo_enable()` functions, it sounds like we need `vim.lsp.config` that follows the model of `vim.diagnostic.config`.

File: runtime/doc/lsp.txt
Comment (p00f) [reply runtime/doc/lsp.txt:702]: I'll move it; I was suggested `buf_`

File: runtime/doc/lsp.txt
Comment (p00f) [reply runtime/doc/lsp.txt:708]: What if we don't add a getter and make it so that it toggles by default if there is no `enable` parameter, and enables/disables if there is?

File: runtime/doc/news.txt
Comment (p00f) [reply runtime/doc/news.txt]: done

File: runtime/lua/vim/lsp/_inlay_hint.lua
Comment (glepnir) [reply runtime/lua/vim/lsp/_inlay_hint.lua:198]: i had point it out `changetick ` increase without text changed and this change will not trigger `on_lines` callback . and there has `on_changedtick` in `nvim_buf_attach` . these problem  i had comment in prev :-)

**Comment by clason** [comment]
I believe the possibility of (easily) toggling diagnostics is something that's been missing from that API as well. I do agree (as I've written before) that it makes sense to look at this (toggling "code overlays") as a general issue.

**Comment by mfussenegger** [comment]
> Does LSP actually require the enable/disable stuff to be configured client-side? Seems like this could/should be a setting that could be sent to the server, using a standard "configure" LSP request. Does LSP have something like that? Can the client-side "capabilities" map be used to enable/disable inlay hints?

Given that inlay hints are requested by the client it's also up to the client to control it. The server has no business controlling client behaviour

> When I suggested skipping the public API in https://github.com/neovim/neovim/pull/23736 , my thinking was: just enable it by default. It doesn't need to be configurable.

Inlay hints was always the first thing I turned off in intellij as it's very intrusive. We can discuss the default but it definitely needs an option to turn it off.  And I'd say not just the display but the requests as well.  Users who don't use it shouldn't need to spend processing power on it.

File: runtime/lua/vim/lsp/_inlay_hint.lua
Comment (p00f) [reply runtime/lua/vim/lsp/_inlay_hint.lua]: done

**Comment by justinmk** [comment]
> An option could be to extend vim.lsp.start/start_client with flags for the different features. E.g. `start({ ..., formatting = false, semantic_tokens = false, inlay_hints = true, ... })`

That sounds like a good compromise since it's already the current pattern. That avoids numerous `foo_enable` functions and gives us time to decide if `vim.lsp.config` is needed/acceptable.

**Comment by clason** [comment]
> That avoids numerous foo_enable functions and gives us time to decide if vim.lsp.config is needed/acceptable.

No. The `foo_enable` functions are about _runtime_ choices, not "all or nothing". These are two separate issues that we should not conflate.

**Comment by justinmk** [comment]
> We have so far only ever considered options 1 and 2, because we got away with it. But this feature crosses a line where option 3 becomes important

With a `vim.start()` parameter it is possible to mimic a "runtime" choice by stopping/restarting the server. In any case that can be a followup, it doesn't need to block the minimal approach, which is what we should start with.

**Comment by clason** [comment]
> With a vim.start() parameter it is possible to mimic a "runtime" choice by stopping/restarting the server. 

Which may require multiple seconds for starting up and re-indexing the project... 

I sincerely believe that "runtime choice" is the primary use case for inlay hints, different from all prior LSP features we added (and would like to point out that this already _is_ the followup).

Anyway, as long as we are clear on what the ask here is, I will hold my peace from now on.

**Comment by justinmk** [comment]
> I sincerely believe that "runtime choice" is the primary use case for inlay hints,

I could be out of touch, but I didn't think that toggling inlay hints on/off at will was a common thing. But again (1) the door is open for that after we decide a long-term interface, (2) doesn't need to block the minimal approach.

**Comment by p00f** [comment]
@justinmk I think vscode has a thing where it shows inlay hints as long as you hold certain keys

Not saying this should be the reference, just addressing
> didn't think that toggling inlay hints on/off at will was a common thing

**Comment by gpanders** [comment]
> > I sincerely believe that "runtime choice" is the primary use case for inlay hints,
> 
> I could be out of touch, but I didn't think that toggling inlay hints on/off at will was a common thing. But again (1) the door is open for that after we decide a long-term interface, (2) doesn't need to block the minimal approach.

FWIW I didn’t think that either until I tried it. And my almost immediate thought was that I wanted to be able to flip them on when needed, but disable them otherwise. Personally, I find them far too noisy to leave enabled all the time.

**Comment by justinmk** [comment]
@mfussenegger 

> A user could change the server capability, to trick the client into thinking it's not supported

But the client also declares its capabilities (`vim.lsp.start({capabilities={...}})`), don't see why a "trick" is needed.

> An option could be to extend vim.lsp.start/start_client with flags for the different features. E.g. `start({ ..., formatting = false, semantic_tokens = false, inlay_hints = true, ... })` 

Let's forget that.  [coc.nvim](https://github.com/neoclide/coc.nvim/commit/2bd87f66c7048e6c7c3a3d067be5c5e6e0036d33) for example, suggests there is a ridiculous amount of configuration people want to fiddle with. Can't jam that all into `vim.lsp.start()`, and see below...

> A `vim.lsp.config` could act as a skeleton, but it would need a way to distinguish between settings for _all_ clients or specific clients.

`vim.lsp.get_active_clients()[1]` shows me:

```
{
  config = {
    _on_attach = <function 3>,
    autostart = true,
    capabilities = { … },
    settings = {
      Lua = {
        diagnostics = {
          globals = { "vim" }
        },
        runtime = { … },
      }
    },
  },
  server_capabilities = { … },
  ...
}
```

Why isn't this the interface for setting client config?

    vim.lsp.get_client().config.settings.inlay_hints = true

I can imagine some cases where "current buf has multiple clients", then we might need: 

    vim.lsp.buf_clients.config.settings.inlay_hints = true

File: runtime/doc/lsp.txt
Comment (justinmk) [reply runtime/doc/lsp.txt:704]: still wondering where the `buf_` suggestion came from..

File: runtime/doc/lsp.txt
Comment (justinmk) [reply runtime/doc/lsp.txt:710]: Anything that results in 1 function instead of 3 is good enough.

Comment (p00f) [reply runtime/doc/lsp.txt]: done

File: runtime/doc/lsp.txt
Comment (p00f) [reply runtime/doc/lsp.txt]: It came from clason

File: runtime/doc/lsp.txt
Comment (p00f) [reply runtime/doc/lsp.txt]: done

File: runtime/doc/lsp.txt
Comment (justinmk) [inline runtime/doc/lsp.txt:1324]: Convention is:

- 0 = current buffer
- `nil` = all buffers (don't necessarily need to implement this here)

Comment (p00f) [reply runtime/doc/lsp.txt:1324]: done

**Review by justinmk — APPROVED** [review]
LGTM (after CI is passing), will wait for other approvals, ping after a few days if not merged already.

Thanks for your persistence and patience!

**Review by lewis6991 — APPROVED** [review]

**Review by mfussenegger — APPROVED** [review]

**Review by mfussenegger — COMMENTED** [review]
Approved this, but I think we should look into doing some API adjustments before this gets released.


Inlay hints have optional `textEdits`, and we could (or should?) allow users to apply them. That will likely need another function. With the current approach this would mean something like `vim.lsp.buf.inlay_hints_run()` (or `apply` or `apply_edits` or whatever). Which would just forward to `_inlay_hints.xy()`.

I'm thinking maybe we could go with a callable module approach similar to how `vim.iter` works. We'd have `vim.lsp.inlay_hints()` to toggle, enable or disable and could add additional functionality as needed in `vim.lsp.inlay_hints.xy()`.
This would remove the indirection/forwarding to a private module, and we could keep stuff properly encapsulated via lua `local`s and it's module system.


The same approach would also work for `codelens`. We could toggle/enable/disable via `vim.lsp.codelens()`, but we still have `vim.lsp.codelens.on_codelens` for the handler and `vim.lsp.codelens.run()` to execute a lens.

File: runtime/doc/lsp.txt
Comment (clason) [reply runtime/doc/lsp.txt]: Would be nice to do this in a follow-up, though, as now you can't just map to `vim.lsp.buf.inlay_hints()` (like you can with most other `vim.lsp.buf` functions).

**Pull request merged**
